_posts/2019-02-27-vue-with-typescript.md

```diff
@@ -0,0 +1,1331 @@
+
+To provide this `$http` member, we have to add the correct plugin to our code.
+In the [awesome-vue project on GitHub](https://github.com/vuejs/awesome-vue){:target="_blank" rel="noopener noreferrer"}, we can find an [extensive list of HTTP plugins](https://github.com/vuejs/awesome-vue#http-requests){:target="_blank" rel="noopener noreferrer"}.
+We will use [axios](https://github.com/axios/axios){:target="_blank" rel="noopener noreferrer"} as our HTTP library but we will use [vue-axios](https://github.com/imcvampire/vue-axios){:target="_blank" rel="noopener noreferrer"} for the bindings with Vue in TypeScript as axios does not provide the necessary typings for axios in Vue.
+
+## Installing a plugin
+We follow the installation instructions for vue-axios which are pretty straightforward:
+
+```text
+$ npm i axios vue-axios
+```
+
+As you noticed we also installed axios.
+This is because vue-axios only focuses on the TypeScript bindings for Vue and does not include the actual axios library.
+Vue-axios basically turns the axios library into a plugin compatible for Vue.
+After that, we have to signal to Vue that we want to use this plugin.
+We add a `Vue.use(plugin, options)` statement in our main.ts so it looks like this:
+
+```typescript
+import axios from 'axios';
+import Vue from 'vue';
+import VueAxios from 'vue-axios';
+import App from './App.vue';
+
+Vue.config.productionTip = false;
+
+Vue.use(VueAxios, axios);
+
+new Vue({
+	render: (h) => h(App),
+}).$mount('#app');
+```
+
+The important part is that we put the `Vue.use(...)` statement before we actually bootstrap the application with `new Vue(...)`.
+
+## The effect of adding a plugin
+So we have added a plugin, but what does that actually mean?
+What is the effect on our Vue code?
+The main effect is that we now have `$http` accessible in every Vue component.
+This means that we can now have this.`$http` in our classes in which a unique instance of the axios library for the whole application will be plugged.
+When we check the [typings](https://github.com/axios/axios/blob/master/index.d.ts){:target="_blank" rel="noopener noreferrer"}  from axios, we find that we now have methods like `get(...)`, `post(...)` and many more default REST methods available in our code through the \$http member in which an instance of axios is present.
+
+Methods like `get(...)` and `post(...)` will also exist on other HTTP libraries that we can add to Vue.
+It is not obliged by Vue to provide these same functionalities in another HTTP library.
+But it makes sense for library creators to comply to the standard set by the HTTP library of preference as chosen by Vue, in this case axios.
+Otherwise it would not be easy to change certain libraries for another one.
+
+## Using the axios plugin for performing HTTP calls
+A Vue component has multiple lifecycle hooks with the most interesting ones for what we want to do: `created()` and `mounted()`.
```

Comment:
Would add a link to https://vuejs.org/v2/guide/instance.html#Instance-Lifecycle-Hooks

A Vue component has multiple lifecycle hooks of which the most interesting ones for what we want to do are `created()` and `mounted()`.